File: December - 25/KishoreMuruganantham_treasure_hunt_locked_maze.R
input <- file("stdin", "r")
lines <- readLines(input)
dims <- as.integer(unlist(strsplit(lines[1], "\\s+")))
r <- dims[1]
c <- dims[2]

grid <- matrix("", nrow = r, ncol = c)
start_pos <- c(0, 0)

for (i in 1:r) {
  row_chars <- strsplit(lines[i + 1], "")[[1]]
  grid[i, ] <- row_chars
  if ("S" %in% row_chars) {
    j <- which(row_chars == "S")
    start_pos <- c(i, j)
  }
}

get_key_bit <- function(ch) {
  utf8ToInt(ch) - utf8ToInt("a")
}
get_door_bit <- function(ch) {
  utf8ToInt(ch) - utf8ToInt("A")
}

visited <- array(FALSE, dim = c(r, c, 1024))

q_r <- integer(2000000)
q_c <- integer(2000000)
q_mask <- integer(2000000)
q_dist <- integer(2000000)

head <- 1
tail <- 1

q_r[tail] <- start_pos[1]
q_c[tail] <- start_pos[2]
q_mask[tail] <- 0
q_dist[tail] <- 0
tail <- tail + 1
visited[start_pos[1], start_pos[2], 1] <- TRUE

found <- FALSE
dr <- c(0, 0, 1, -1)
dc <- c(1, -1, 0, 0)

while (head < tail) {
  cr <- q_r[head]
  cc <- q_c[head]
  cmask <- q_mask[head]
  d <- q_dist[head]
  head <- head + 1
  
  if (grid[cr, cc] == "T") {
    cat(d, "\n", sep = "")
    found <- TRUE
    break
  }
  
  for (k in 1:4) {
    nr <- cr + dr[k]
    nc <- cc + dc[k]
    
    if (nr >= 1 && nr <= r && nc >= 1 && nc <= c) {
      cell <- grid[nr, nc]
      if (cell == "#") next
      
      nmask <- cmask
      allowed <- TRUE
      
      if (cell >= "a" && cell <= "j") {
        nmask <- bitwOr(cmask, bitwShiftL(1, get_key_bit(cell)))
      } else if (cell >= "A" && cell <= "J") {
        req <- get_door_bit(cell)
        if (bitwAnd(cmask, bitwShiftL(1, req)) == 0) {
          allowed <- FALSE
        }
      }
      
      if (allowed && !visited[nr, nc, nmask + 1]) {
        visited[nr, nc, nmask + 1] <- TRUE
        q_r[tail] <- nr
        q_c[tail] <- nc
        q_mask[tail] <- nmask
        q_dist[tail] <- d + 1
        tail <- tail + 1
      }
    }
  }
}

if (!found) {
  cat("-1\n")
}

File: December - 25/KishoreMuruganantham_treasure_hunt_locked_maze.swift
import Foundation
var lines = [String]()
while let line = readLine() { lines.append(line) }
if !lines.isEmpty {
    let dims = lines[0].split(separator: " ").compactMap { Int($0) }
    let r = dims[0]
    let c = dims[1]
    
    var grid = [[Character]]()
    var startPos = (0, 0)
    
    for i in 1...r {
        let row = Array(lines[i])
        grid.append(row)
        if let idx = row.firstIndex(of: "S") {
            startPos = (i - 1, idx)
        }
    }
    
    func getKeyBit(_ char: Character) -> Int {
        return Int(char.asciiValue! - Character("a").asciiValue!)
    }
    func getDoorBit(_ char: Character) -> Int {
        return Int(char.asciiValue! - Character("A").asciiValue!)
    }
    
    var visited = Set<Int>() 
    
    struct State {
        let r: Int, c: Int, mask: Int, dist: Int
    }
    
    var q = [State]()
    let startState = State(r: startPos.0, c: startPos.1, mask: 0, dist: 0)
    q.append(startState)
    visited.insert((startPos.0 << 15) | (startPos.1 << 10) | 0)
    
    var head = 0
    let dr = [0, 0, 1, -1]
    let dc = [1, -1, 0, 0]
    var found = false
    
    while head < q.count {
        let curr = q[head]
        head += 1
        
        if grid[curr.r][curr.c] == "T" {
            print(curr.dist)
            found = true
            break
        }
        
        for k in 0..<4 {
            let nr = curr.r + dr[k]
            let nc = curr.c + dc[k]
            
            if nr >= 0 && nr < r && nc >= 0 && nc < c {
                let cell = grid[nr][nc]
                if cell == "#" { continue }
                
                var nmask = curr.mask
                var allowed = true
                
                if cell >= "a" && cell <= "j" {
                    nmask |= (1 << getKeyBit(cell))
                } else if cell >= "A" && cell <= "J" {
                    if (nmask & (1 << getDoorBit(cell))) == 0 {
                        allowed = false
                    }
                }
                
                if allowed {
                    let hash = (nr << 15) | (nc << 10) | nmask
                    if !visited.contains(hash) {
                        visited.insert(hash)
                        q.append(State(r: nr, c: nc, mask: nmask, dist: curr.dist + 1))
                    }
                }
            }
        }
    }
    
    if !found {
        print("-1")
    }
}
